loader/loader.go

```diff
@@ -114,43 +141,148 @@ func (fl fileLoader) All(paths []string) (*Result, error) {
 // paths while applying the given filters. If any filter returns true, the
 // file/directory is excluded.
 func (fl fileLoader) Filtered(paths []string, filter Filter) (*Result, error) {
-	return all(paths, filter, func(curr *Result, path string, depth int) error {
+	root := newResult()
+	errors := Errors{}
+	var signatures *bundle.SignaturesConfig
+
+	err := all(paths, filter, root, func(curr *Result, path string, depth int) error {
+
+		// check for the signatures file
+		if !fl.skipVerify && strings.HasSuffix(path, bundle.SignaturesFile) {
+			bs, err := ioutil.ReadFile(path)
+			if err != nil {
+				return err
+			}
+
+			if err := util.UnmarshalJSON(bs, &signatures); err != nil {
+				return fmt.Errorf("bundle load failed on signatures decode")
+			}
+		} else if !strings.HasSuffix(path, bundle.SignaturesFile) {
+			// generate the normalized relative path for the file using the depth of the directory structure
+			// ex. path: "bundle/.manifest", depth: 1, rel path: ".manifest"
+			// ex. path: "bundle/http/example/authz/authz.rego", depth: 4, rel path: "http/example/authz/authz.rego"
+
+			parts := strings.Split(path, "/")
+			if depth > len(parts) {
+				return fmt.Errorf("incorrect directory hierarchy for path %v", path)
+			}
+			parts = parts[len(parts)-depth:]
+
+			relPath := strings.Join(parts, "/")
+			if relPath == "" {
+				// file located at root of the directory
+				relPath = filepath.Base(path)
+			}
+
+			fl.descriptors = append(fl.descriptors, &descriptor{
+				result:  curr,
+				path:    path,
+				relPath: relPath,
+				depth:   depth,
+			})
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if signatures != nil {
+		root.Signatures = signatures
+
+		if fl.bvc == nil {
+			return nil, fmt.Errorf("missing configuration to verify bundle signature")
+		}
```

Comment:
What happens if someone specifies a directory with >1 bundle in it? Say I had:
```
./bundles/b1
./bundles/b1/.signatures.json
./bundles/b1/policy.rego
./bundles/b2
./bundles/b2/.signatures.json
./bundles/b1/other-policy.rego
```
And I do something like `opa run ./bundles` IIRC both bundles get handled by a single call to `Filtered()`. If I'm reading this correctly we would only verify the last bundle signature file we found, right?

We could potentially say if someone wants bundle verification to happen they need to use the `-b` option and specify them as bundles. Not sure how others feel about that.. but we did want to push more people to using that. I'd almost prefer having this error out saying "found a signed bundle, use --bundle instead" rather than potentially giving wrong results _or_ having a ton of complexity in the `--data` loading path.

Just kind of brainstorming here.. if we do go the "unsupported" path we could iterate later on to add support as needed.

Reply:
> What happens if someone specifies a directory with >1 bundle in it? Say I had:

With the current implementation, I'd expect an error since we do not support multiple JWTs.

Reply:
> I'd expect an error since we do not support multiple JWTs

Not sure I follow, if we load those files as if they are bundles, eg:

```
./bundles/b1
./bundles/b1/.signatures.json
./bundles/b1/policy.rego
./bundles/b2
./bundles/b2/.signatures.json
./bundles/b1/other-policy.rego
```
should be loaded the same as

```
./bundles/b1.tar.gz
./bundles/b2.tar.gz
```

if I did `opa run ./bundles`. Each bundle still has a single JWT that should verify its content. I guess the question is whether or not we want the non-bundle-mode loading option to even try to handle these or if we say a user has to specify `opa run -b ./bundles/b1 ./bundles/b2` or `opa run -b ./bundles/b1.tar.gz ./bundles/b2.tar.gz` instead if they want validation.

Reply:
>Not sure I follow, if we load those files as if they are bundles, eg:

If we're passing multiple paths and interpreting them as bundles (e.g., `-b` was given) then yes, but I thought the point was that one of those was not happening.

With the rest of this PR it looks like we're erring on the side of security; if the user does `opa run .` without `-b` and the filesystem contains a signatures.json file, it will be honored. This makes sense to some extent though it does depart from the rest of the file loading conventions.

Thinking about this a bit more, I'm a bit surprised we pickup `.signatures.json` files under b1 and b2 when the load path is `.` or `bundles/`. I was under the impression the signatures.json file could only exist at the root of the bundle. This doesn't matter too much though since users could also do `opa run ./bundles/b1 ./bundles/b2` (without `-b`) and in that case the signatures files would get honored.

Reply:
Closing the loop on the offline discussion. The plan is to leave non-bundle-mode loading alone for now. OPA will not verify signatures unless bundle mode is enabled. This simplifies the implementation, removes room for error, and avoids the backwards incompatibility issue of making certain files have meaning when non-bundle-mode loading is in use.

Reply:
Done. Bundle signing and verification is supported only in the bundle mode.